cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,99 @@
+ */
+class KafkaQuorumCheck {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaQuorumCheck.class.getName());
+    private final Admin ac;
+    private final Reconciliation reconciliation;
+    private final Future<QuorumInfo> quorumInfoFuture;
+    private final String controllerQuorumFetchTimeoutMs;
+
+    KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, String controllerQuorumFetchTimeoutMs) {
+        this.ac = ac;
+        this.reconciliation = reconciliation;
+        this.quorumInfoFuture = describeMetadataQuorum();
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Determine whether the given controller can be rolled without affecting the quorum
+     */
+    Future<Boolean> canRoll(int podId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        return canRollController(quorumInfoFuture, podId);
+    }
+
+    private Future<Boolean> canRollController(Future<QuorumInfo> quorumInfo, int podId) {
+        return quorumInfo.map(info -> {
+            int leaderId = info.leaderId();
+            Map<Integer, Long> replicaIdToLastCaughtUpTimestampMap = info.voters().stream().collect(Collectors.toMap(
+                    QuorumInfo.ReplicaState::replicaId,
+                    replicaState -> replicaState.lastCaughtUpTimestamp().isPresent() ? replicaState.lastCaughtUpTimestamp().getAsLong() : -1));
+            boolean canRoll = wouldAffectQuorum(leaderId, replicaIdToLastCaughtUpTimestampMap);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean wouldAffectQuorum(int leaderId, Map<Integer, Long> replicaStates) {
+        int totalNumVoters = replicaStates.size();
+        AtomicInteger numOfCaughtUpVoters = new AtomicInteger();
+        long leaderLastCaughtUpTimestamp = replicaStates.get(leaderId);
+        if (leaderLastCaughtUpTimestamp < 0) {
+            LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the leader replica {} ", leaderId);
+            return false;
+        }
+        LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for the leader replica {} is {}", leaderId, leaderLastCaughtUpTimestamp);
+        replicaStates.forEach((replicaId, lastCaughtUpTimestamp) -> {
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
+            } else {
+                if (leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp < Long.parseLong(controllerQuorumFetchTimeoutMs)) {
```

Comment:
We are parsing the same string multiple times here.
I would declare `controllerQuorumFetchTimeoutMs` as a Long and parsing the received String in the constructor just one time (instead of assigning the String) or I would change the constructor to get a Long and parsing the String value in the KafkaRoller code (where the `CONTROLLER_QUORUM_FETCH_TIMEOUT_MS_CONFIG_NAME` is read and assigned to the restartContext).

Reply:
good point!